File: triangle/triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        
        m = len(triangle)
        
        
        if not triangle:
            return 0
        if len(triangle) == 1:
            return triangle[0][0]
    
    
        for i in range(m-2, -1, -1):
            
            for j in range(len(triangle[i])):
                
                triangle[i][j] += min(triangle[i+1][j], triangle[i+1][j+1])
                
        return triangle[0][0]
            
            

File: triangle/triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
     
        
        
        int t_len = triangle.size();
        if(t_len == 0){
            return 0;
        }
        if(t_len == 1){
            return triangle.get(0).get(0); 
        }
        
        List<Integer> prev = triangle.get(t_len - 1);
        
        for(int i = t_len-2; i>= 0; i--){
            List<Integer> curr = new ArrayList<Integer>();
            
            for (int j = 0; j < triangle.get(i).size(); j++){
                curr.add(Math.min((prev.get(j) + triangle.get(i).get(j)), (prev.get(j+1) + triangle.get(i).get(j)) ));
            }
            
            prev = curr;
            
        }
        
        return prev.get(0);
        
    }
}